homeassistant/components/sensor/dsmr.py

```diff
@@ -84,122 +88,48 @@ def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
-    reader = hass.loop.create_task(dsmr.dsmr_parser.read(queue))
-
-    # serial telegram reader is a infinite looping task, it will only resolve
-    # when it has an exception, in that case log this.
-    def handle_error(future):
-        """If result is an exception log it."""
-        _LOGGER.error('error during initialization of DSMR serial reader: %s',
-                      future.exception())
-    reader.add_done_callback(handle_error)
-
-    # add task to receive telegrams and update entities
-    hass.async_add_job(dsmr.read_telegrams)
-
-
-class DSMR:
-    """DSMR interface."""
-
-    def __init__(self, hass, config, devices, queue):
-        """Setup DSMR serial interface, initialize, add device entity list."""
-        from dsmr_parser.serial import (
-            SERIAL_SETTINGS_V4,
-            SERIAL_SETTINGS_V2_2,
-            AsyncSerialReader
-        )
-        from dsmr_parser import telegram_specifications
-
-        # map dsmr version to settings
-        dsmr_versions = {
-            '4': (SERIAL_SETTINGS_V4, telegram_specifications.V4),
-            '2.2': (SERIAL_SETTINGS_V2_2, telegram_specifications.V2_2),
-        }
-
-        # initialize asynchronous telegram reader
-        dsmr_version = config[CONF_DSMR_VERSION]
-        self.dsmr_parser = AsyncSerialReader(
-            device=config[CONF_PORT],
-            serial_settings=dsmr_versions[dsmr_version][0],
-            telegram_specification=dsmr_versions[dsmr_version][1],
-        )
-
-        # keep list of device entities to update
-        self.devices = devices
-
-        self._queue = queue
-
-        # initialize empty telegram
-        self._telegram = {}
-
-        # forward stop event to reading loop
-        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP,
-                                   self._queue.put_nowait)
-
-    @asyncio.coroutine
-    def read_telegrams(self):
-        """Receive parsed telegram from DSMR reader, update entities."""
-        while True:
-            # asynchronously get latest telegram when it arrives
-            event = yield from self._queue.get()
-
-            # stop loop if stop event was received
-            if getattr(event, 'event_type', None) == EVENT_HOMEASSISTANT_STOP:
-                self._queue.task_done()
-                return
-
-            self._telegram = event
-            _LOGGER.debug('received DSMR telegram')
-
-            # make all device entities aware of new telegram
-            for device in self.devices:
-                yield from device.async_update_ha_state()
+    def update_entities_telegram(telegram):
+        """Updates entities with latests telegram & trigger state update."""
+        # make all device entities aware of new telegram
+        for device in devices:
+            device.telegram = telegram
+            hass.async_add_job(device.async_update_ha_state)
+            # hass.loop.create_task(device.async_update_ha_state())
```

Comment:
May I suggest adding: https://github.com/aequitas/pytest-eradicate to the project?